### PR TITLE
[Reliability] Update data table to use React

### PR DESCRIPTION
**Pull request by alisterdev**
- [x] Add react data table
- [x] Fix sorting by reliable column (Redmine 10204)

### CODE REVIEW

**Comment by kongtiaowang** [comment]
I think we do the same task. 
That is my PR. https://github.com/aces/Loris/pull/2144

**Comment by alisterdev** [comment]
@driusan Do you know why  `codecov/patch` fails and how can I fix it?

**Comment by driusan** [comment]
It shows up as failing if the testing coverage doesn't go up, I think, but it isn't required (or at least shouldn't be).

I think the reason it's showing up as unmergable is because there's no review (which is required), not because of codecov (which isn't)

**Comment by alisterdev** [comment]
@driusan Do you want to code review at merge it then? @stellarxo already tested it

**Comment by alisterdev** [comment]
@driusan Ready to be merged

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc [hunk 1]
```diff
@@ -69,32 +69,37 @@ class NDB_Menu_Filter_reliability extends NDB_Menu_Filter
...
         // only view their own profiles, unless they have permission to see all
         if (!($user->hasPermission('reliability_edit_all') || $user->hasPermission('access_all_profiles'))) {
-            $query .= " AND candidate.centerID = '" . $user->getData('CenterID') . "' "; 
+            $query .= " AND candidate.centerID = '" . $user->getData('CenterID') . "' ";
         }
         $config=&NDB_Config::singleton();
         $useProjects = $config->getSetting("useProjects");
         // set the class variables
-        $this->columns = array('candidate.PSCID', 'reliability.reliability_center_id AS Site', 'session.SubprojectID as Cohort', 'session.CandID AS DCCID', 'candidate.Gender', 'session.Visit_label');
+        $this->columns = array('candidate.PSCID', 'reliability.reliability_center_id', '(SELECT name FROM psc WHERE CenterID=reliability.reliability_center_id) as site', 'session.SubprojectID as Cohort', 'session.CandID AS DCCID', 'candidate.Gender', 'session.Visit_label');
```
Comment (driusan) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: Why is this changed to a subselect? Shouldn't it be a join?

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc [hunk 2]
```diff
@@ -69,32 +69,37 @@ class NDB_Menu_Filter_reliability extends NDB_Menu_Filter
...
         if ($useProjects === "true") {
-            $this->columns[] = 'candidate.ProjectID as Project';
+            $this->columns[] = '(SELECT Name FROM Project WHERE ProjectID=candidate.ProjectID) as Project';
```
Comment (driusan) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: This looks like it should also be done by joining the table instead of using a subselect..

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc [hunk 3]
```diff
@@ -507,11 +444,43 @@ class NDB_Menu_Filter_reliability extends NDB_Menu_Filter
...
                'Instrument'=>$params['Instrument'],
-               'Manual_Swap'=>'no') 
+               'Manual_Swap'=>'no')
         );
         return array("message" => "Inserted new reliability candidate".$params['PSCID']."/".$params['Visit_label'] );
     }
-    
+
+    /**
+     * Override toJSON to append a calculated field to the standard output
```
Comment (driusan) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: I find the usage of the phrase "standard output" a little confusing, since it kind of sounds like a reference to STDOUT.

File: modules/reliability/templates/menu_reliability.tpl [hunk 4]
```diff
@@ -1,96 +1,93 @@
...
 
+                  <div class="row">
+                      <div class="col-sm-4 col-md-3 col-xs-12 col-md-offset-5 col-sm-offset-4">
+                          <input type="submit" name="filter" value="Show Data" class="btn btn-sm btn-primary col-xs-12" />
+                      </div>
 
+                      <div class="visible-xs col-xs-12"> </div>
+                      <div class="visible-xs col-xs-12"> </div>
+                      <div class="visible-xs col-xs-12"> </div>
+                      <div class="visible-xs col-xs-12"> </div>
```
Comment (driusan) [inline modules/reliability/templates/menu_reliability.tpl]: Didn't you make a React table filter component so that the data table and filters can be connected?

Would it make sense to use that here instead of smarty?

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc (hunk 1 above)
Comment (alisterdev) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: @driusan  It is changed because in the old version the conversion between numeric CenterID and the actual name was done on the front-end. Now, I get this information in the query right away. 

Don't know how a JOIN would benefit here.

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc (hunk 2 above)
Comment (alisterdev) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: The join was not used before, is there any benefit of joining the whole table for a couple of values?

File: modules/reliability/templates/menu_reliability.tpl (hunk 4 above)
Comment (alisterdev) [inline modules/reliability/templates/menu_reliability.tpl]: It would, but this particular update only addresses the bug described in the redmine ticket.
Implementing react data table was just the quickest solution for that.

**Comment by alisterdev** [comment]
@driusan Replied to comments

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc (hunk 1 above)
Comment (driusan) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: JOINs are significantly faster, because subselects need to executed once per row, while SQL databases are built around/optimized for JOINs/relational algebra, and only need to execute once per query.

They also break less often, because JOINs don't kill the whole query if something changes causing the query to return multiple rows.

Comment (alisterdev) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: @driusan Updated to use joins

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc [hunk 5]
```diff
@@ -69,32 +69,41 @@ class NDB_Menu_Filter_reliability extends NDB_Menu_Filter
         foreach($pscRows AS $row) {
             $this->centerIDMap[$row['CenterID']] = $row['Name'];
         }
-        
+
         $this->_populate_reliability_table();
 
-        $query = " FROM candidate, session, flag, reliability WHERE reliability.CommentID=flag.CommentID AND candidate.CandID=session.CandID AND session.ID=flag.SessionID";
+        $query = " FROM session, flag, " .
+            "reliability LEFT JOIN psc p ON reliability.reliability_center_id=p.CenterID, " .
```
Comment (driusan) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: I think there's a syntax error here (comma at the end of the line in the ON clause.)

As long as you're updating that, can you put brackets around the ON clause for readability? And it might be a good idea to make the query consistently ANSI join syntax ("LEFT JOIN ... ON") instead of half cross joins (comma separating tables w/ where clause) and half ANSI joins.

File: modules/reliability/templates/menu_reliability.tpl (hunk 4 above)
Comment (driusan) [inline modules/reliability/templates/menu_reliability.tpl]: Okay, it just looked like you were changing the filter portion anyways, so I wasn't sure why you didn't go all out. It might just look that way because of spacing..

Comment (alisterdev) [inline modules/reliability/templates/menu_reliability.tpl]: Yeh, I didn't touch the filter portion, it's just indentation

File: modules/reliability/php/NDB_Menu_Filter_reliability.class.inc (hunk 5 above)
Comment (alisterdev) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: Comma is at the end of the line because there is another table with another join following. So it goes like

`SELECT .... FROM table1, table2, (table3 JOIN ... ON ....), (table4 JOIN ... ON ....) WHERE ....`

Comment (alisterdev) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: @driusan I'm not an expert in SQL standards so I'm not exactly sure how you want me to reformat that query

Comment (alisterdev) [inline modules/reliability/php/NDB_Menu_Filter_reliability.class.inc]: fixed

**Comment by alisterdev** [comment]
@driusan Addresses your query comments, please review/merge

File: modules/reliability/jsx/columnFormatter.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  if (column === 'Cohort') {
+    var cohort = row['Cohort'];
+    if (cohort === "1") {
+      return <td>6 month</td>;
+    } else if (cohort === "2") {
+      return <td>12 month</td>;
+    } else if (cohort === "3") {
+      return <td>Control</td>;
```
Comment (driusan) [inline modules/reliability/jsx/columnFormatter.js]: These are hardcoded IBIS values. Can you change this to an AJAX call (or React prop) that wraps around Utility::getSubprojectList()?

(Or even better long term would be to find a way to put the subprojects into the `loris.` js helper, but that's probably too much work for this release.)

Comment (alisterdev) [inline modules/reliability/jsx/columnFormatter.js]: @driusan Fixed to pull values from `Utility::getSubprojectList()` and rebased
